flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskFinalCheckpointsTest.java

```diff
@@ -840,6 +846,72 @@ public void testOperatorSkipLifeCycleIfFinishedOnRestore() throws Exception {
         }
     }
 
+    /**
+     * This test verifies for tasks that finished on restore, when taking unaligned checkpoint the
+     * asynchronous part would wait for the channel states futures get completed, which means the
+     * barriers are aligned.
+     */
+    @Test
+    public void testWaitingForUnalignedChannelStatesIfFinishedOnRestore() throws Exception {
+        OperatorID operatorId = new OperatorID();
+        try (StreamTaskMailboxTestHarness<String> harness =
+                new StreamTaskMailboxTestHarnessBuilder<>(
+                                OneInputStreamTask::new, BasicTypeInfo.STRING_TYPE_INFO)
+                        .modifyStreamConfig(
+                                streamConfig -> streamConfig.setUnalignedCheckpointsEnabled(true))
+                        .addInput(BasicTypeInfo.STRING_TYPE_INFO, 3)
+                        .setCollectNetworkEvents()
+                        .setTaskStateSnapshot(1, TaskStateSnapshot.FINISHED_ON_RESTORE)
+                        .setupOperatorChain(new TestFinishedOnRestoreStreamOperator())
+                        .chain(
+                                operatorId,
+                                new TestFinishedOnRestoreStreamOperator(operatorId),
+                                StringSerializer.INSTANCE)
+                        .finish()
+                        .build()) {
+            // Finish the restore, including state initialization and open.
+            harness.processAll();
+
+            TestCheckpointResponder checkpointResponder = harness.getCheckpointResponder();
+            checkpointResponder.setAcknowledgeLatch(new OneShotLatch());
+            checkpointResponder.setDeclinedLatch(new OneShotLatch());
+
+            CheckpointBarrier unalignedBarrier =
+                    new CheckpointBarrier(2, 2, CheckpointOptions.unaligned(getDefault()));
+
+            // On first unaligned barrier, the task would take snapshot and start the asynchronous
+            // part. We slightly extend the process to make the asynchronous part start executing
+            // before the other barriers arrived.
+            harness.processEvent(unalignedBarrier, 0, 0);
+            Thread.sleep(1000);
```

Comment:
maybe extract the `500` constant used in other unit tests in this file and use the same constant here as well?